### PR TITLE
Plot interaction methods

### LINKED ISSUE #734: Plot: Get mouse position
To use plots interactively, we need to have a way of retrieving the mouse position in plot coordinates.

I'm not sure yet how to do this but can think of two ways that could work.

**1.)** Store the mouse position from the last frame in memory and use a specialized function to retrieve it before drawing the plot again.
```rust
let mut position = Pos2::ZERO;
let plot = Plot::new("plot_name")
    .get_mouse_position(ui, &mut position)
    .text(Text::new(position, "mouse was here"));
ui.add(plot)
```
Pro: Simple
Con: The mouse position always lags behind by one frame.

**2.)** Completely rework how plots are created, using a closure to add the contents.
This idea is based on how [implot-rs](https://github.com/4bb4/implot-rs) works.
```rust
Plot::new("plot_name")
    .build(ui, |plot_ui| {
        let position = plot_ui.mouse_position();
        plot_ui.text(Text::new(position, "mouse IS here"))
    });
```
Pro: Mouse position is available the frame it is needed.
Con: Requires many changes. Not sure if it's going to work.

If anyone has another idea, please let me know.

### PR DESCRIPTION
**Pull request by EmbersArc**
Closes #734.

I went with option 2 described in #734. This is a small breaking change and updating shouldn't be difficult. In the long run we can use this to add even more methods for interacting with the plot.

Might be a bit rough around the edges so feel free to suggest improvements.

### CODE REVIEW

**Review by emilk — CHANGES_REQUESTED** [review]
Overall this looks great!

What do you think about using `Value` instead of `Pos2` for the values in plot-coordinates?

We could add helpers to `Value` (`to_pos2` and `to_vec2`) instead.

`Value` could also do with a better name, but that's another problem :)

File: egui/src/widgets/plot/mod.rs [hunk 1]
```diff
@@ -534,15 +427,150 @@ impl Widget for Plot {
     }
 }
 
-struct Prepared {
+pub struct PlotUi {
```
[the remark this answers is not in the document]
Comment (EmbersArc) [reply egui/src/widgets/plot/mod.rs:430]: I added the comment except for the note that it should not be created directly, as that is not possible anyway.

**Comment by EmbersArc** [comment]
I had completely missed your review and only just saw it. Sorry about the delay. I agree with all of your suggestions and have made the changes.

**Comment by EmbersArc** [comment]
I tested this today and am pretty happy with how it works. I'd like to add the following features as well though:

* Access to `Ui` inside the closure as well. Either through the `plot_ui` struct or as a second argument.
* A `plot_to_screen` conversion, to enable custom drawing in screen coordinates based on items in the plot.
* A way to get the current plot bounds.

**Review by emilk — APPROVED** [review]
Excellent!

**Pull request merged**